src/linked_hash_map.rs

```diff
@@ -1677,6 +1749,173 @@ impl<'a, K, V> Drop for Drain<'a, K, V> {
+            let i_entry = self
+                .table
+                .find_entry(hash, |o| (*o).as_ref().key_ref().eq(key));
+
+            match i_entry {
+                Ok(occupied) => {
+                    let at = *occupied.get();
+                    self.muv(|| at);
+                    true
+                }
+                Err(_) => false,
+            }
+        }
+    }
+
+    // Updates the pointer to the current element to the one returned by the at closure function.
+    #[inline]
+    fn muv(&mut self, at: impl FnOnce() -> NonNull<Node<K, V>>) {
+        self.cur = at().as_ptr();
+    }
+
+    /// Inserts the provided key and value before the current item. It checks if an entry
+    /// with the given key exists and, if so, replaces its value with the provided `1`
+    /// parameter from the given tuple. The key is not updated; this matters for types that
+    /// can be `==` without being identical.
+    ///
+    /// If the entry doesn't exist, it creates a new one. If a value has been updated, the
+    /// function returns the *old* value wrapped with `Some`  and `None` otherwise.
+    #[inline]
+    pub fn insert_before(&mut self, tuple: (K, V)) -> Option<V>
+    where
+        K: Eq + Hash,
+        S: BuildHasher,
+    {
+        let before = unsafe { NonNull::new_unchecked(self.cur) };
+        self.insert(tuple, || before)
+    }
+
+    /// Inserts the provided key and value after the current item. It checks if an entry
+    /// with the given key exists and, if so, replaces its value with the provided `1`
+    /// parameter from the given tuple. The key is not updated; this matters for types that
+    /// can be `==` without being identical.
+    ///
+    /// If the entry doesn't exist, it creates a new one. If a value has been updated, the
+    /// function returns the *old* value wrapped with `Some`  and `None` otherwise.
+    #[inline]
+    pub fn insert_after(&mut self, tuple: (K, V)) -> Option<V>
+    where
+        K: Eq + Hash,
+        S: BuildHasher,
+    {
+        let before = unsafe { (*self.cur).links.value.next };
+        self.insert(tuple, || before)
+    }
+
+    // Inserts `item` immediately before the element returned by the `before` closure function.
+    #[inline]
+    fn insert(&mut self, tuple: (K, V), before: impl FnOnce() -> NonNull<Node<K, V>>) -> Option<V>
+    where
+        K: Eq + Hash,
+        S: BuildHasher,
+    {
+        let (key, value) = tuple;
+        unsafe {
+            let hash = hash_key(self.hash_builder, &key);
+            let i_entry = self
+                .table
+                .find_entry(hash, |o| (*o).as_ref().key_ref().eq(&key));
+
+            match i_entry {
+                Ok(occupied) => {
+                    let mut node = *occupied.into_mut();
+                    let pv = mem::replace(&mut node.as_mut().entry_mut().1, value);
+                    detach_node(node);
+                    attach_before(node, before());
```

Comment:
I think this is incorrect in the case where you call e.g. `insert_before` on a cursor where the provided key is the same as the element that the cursor is currently pointing to.

So for example, if you have the map `[(2, 2), (3, 3), (4, 4)]`, and the cursor is pointing to the `(3, 3)` entry, and you call `cursor.insert_before((3, 5))`, both `node`  and `before()` here would be the same node and this would result in a corrupted linked list.

We should add a test for this as well.

Reply:
Great catch, thanks! Please refer to my comment about lazy evaluation here https://github.com/kyren/hashlink/pull/25#discussion_r1560852763. As I mentioned, I think I implemented it incorrectly, just because it's not lazy evaluation technically, but must be, to address this particular concern. I will add test case for this scenario.

Good spot, thank you! Please check out my comment on lazy evaluation here: https://github.com/kyren/hashlink/pull/25#discussion_r1560852763. I believe my implementation might be incorrect as it doesn't technically provides lazy evaluation, which is necessary to address this issue. I'll be adding a test case for this scenario.

Reply:
Good spot, thank you! Please check out my comment on lazy evaluation here: https://github.com/kyren/hashlink/pull/25#discussion_r1560852763. I believe my implementation might be incorrect as it doesn't technically provides lazy evaluation, which is necessary to address this issue. I'll be adding a test case for this scenario.

Reply:
I don't actually see how lazy evaluation comes into it at all, if you call `detach_node(node)` and the `before` node is the same node, you don't have a direct way to insert the node back to where it was because it's been removed from the linked list. I was imagining maybe just a check for equality with `before`?

Reply:
You're absolutely correct. My apologies for the oversight - the lazy evaluation logic is indeed unrelated to the corrupted linked-list issue.